# DataFrames

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [1]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [2]:
# help(pd.DataFrame)

In [7]:
# Make sure the seed is in the same cell as the random call
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
import numpy as np 
import pandas as pd
np.random.seed(101)
myrot = np.arange(0,100,2).reshape(2,25)
mydata = np.random.randint(0,101,36).reshape(6,6)

In [4]:
mydata

array([[95, 11, 81, 70, 63, 87],
       [75,  9, 77, 40,  4, 63],
       [40, 60, 92, 64,  5, 12],
       [93, 40, 49, 83,  8, 29],
       [59, 34, 44, 72, 19, 10],
       [76, 95, 87,  0, 73,  8]])

In [8]:
myrot

array([[ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30,
        32, 34, 36, 38, 40, 42, 44, 46, 48],
       [50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80,
        82, 84, 86, 88, 90, 92, 94, 96, 98]])

In [9]:
myindex = ['CA','NY','AZ','TX','RE','JK']

In [10]:
mycolumns = ['Jan','Feb','Mar','Apr','Jun','may']

In [11]:
df = pd.DataFrame(mydata,myindex,mycolumns)
df

Jan  Feb  Mar  Apr  Jun  may
CA   95   11   81   70   63   87
NY   75    9   77   40    4   63
AZ   40   60   92   64    5   12
TX   93   40   49   83    8   29
RE   59   34   44   72   19   10
JK   76   95   87    0   73    8

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6 entries, CA to JK
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     6 non-null      int32
 1   Feb     6 non-null      int32
 2   Mar     6 non-null      int32
 3   Apr     6 non-null      int32
 4   Jun     6 non-null      int32
 5   may     6 non-null      int32
dtypes: int32(6)
memory usage: 192.0+ bytes


In [13]:
df.describe()

Jan       Feb        Mar        Apr        Jun        may
count   6.000000   6.00000   6.000000   6.000000   6.000000   6.000000
mean   73.000000  41.50000  71.666667  54.833333  28.666667  34.833333
std    20.890189  32.40216  20.215506  30.426414  31.091264  32.884140
min    40.000000   9.00000  44.000000   0.000000   4.000000   8.000000
25%    63.000000  16.75000  56.000000  46.000000   5.750000  10.500000
50%    75.500000  37.00000  79.000000  67.000000  13.500000  20.500000
75%    88.750000  55.00000  85.500000  71.500000  52.000000  54.500000
max    95.000000  95.00000  92.000000  83.000000  73.000000  87.000000

In [14]:
df.value_counts()

Jan  Feb  Mar  Apr  Jun  may
40   60   92   64   5    12     1
59   34   44   72   19   10     1
75   9    77   40   4    63     1
76   95   87   0    73   8      1
93   40   49   83   8    29     1
95   11   81   70   63   87     1
Name: count, dtype: int64

In [15]:
df.aggregate

<bound method DataFrame.aggregate of     Jan  Feb  Mar  Apr  Jun  may
CA   95   11   81   70   63   87
NY   75    9   77   40    4   63
AZ   40   60   92   64    5   12
TX   93   40   49   83    8   29
RE   59   34   44   72   19   10
JK   76   95   87    0   73    8>

In [16]:
np.round(df.describe(),2)

Jan    Feb    Mar    Apr    Jun    may
count   6.00   6.00   6.00   6.00   6.00   6.00
mean   73.00  41.50  71.67  54.83  28.67  34.83
std    20.89  32.40  20.22  30.43  31.09  32.88
min    40.00   9.00  44.00   0.00   4.00   8.00
25%    63.00  16.75  56.00  46.00   5.75  10.50
50%    75.50  37.00  79.00  67.00  13.50  20.50
75%    88.75  55.00  85.50  71.50  52.00  54.50
max    95.00  95.00  92.00  83.00  73.00  87.00


# Reading a .csv file for a DataFrame

----

## NOTE: We will go over all kinds of data inputs and outputs (.html, .csv, .xlxs , etc...) later on in the course! For now we just need to read in a simple .csv file.

----

## CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [17]:
pwd

'C:\\Users\\hp\\Documents\\python\\Python Scripts\\venv\\Scripts\\__pycache__\\Machine Learning\\03-Pandas'

#### List the files in your current directory with ls

In [18]:
ls

 Volume in drive C has no label.
 Volume Serial Number is 4C79-1E8C

 Directory of C:\Users\hp\Documents\python\Python Scripts\venv\Scripts\__pycache__\Machine Learning\03-Pandas

14-07-2023  19:19    <DIR>          .
14-07-2023  19:19    <DIR>          ..
07-06-2023  16:08    <DIR>          .ipynb_checkpoints
07-06-2023  16:08    <DIR>          __pycache__
14-07-2023  19:01           702,311 00-Series.ipynb
14-07-2023  19:19           158,129 01-DataFrames.ipynb
14-07-2023  13:11            64,193 02-Conditional-Filtering.ipynb
14-07-2023  18:55           200,461 03-Useful-Methods.ipynb
10-06-2023  19:40         1,017,783 04-Missing-Data.ipynb
16-06-2023  14:54           217,157 05-Groupby-Operations-and-MultiIndex.ipynb
10-06-2023  21:48            65,138 06-Combining-DataFrames.ipynb
10-06-2023  21:54            35,765 07-Text-Methods.ipynb
10-06-2023  22:07            96,548 08-Time-Methods.ipynb
10-06-2023  22:10            93,184 09-Inputs-and-Outputs.ipynb
16-06-2023  15:49     

In [38]:
import pandas as pd
import numpy as np

In [39]:
df = pd.read_csv('tips.csv')

In [40]:
df.info()
df.sample(n=2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


total_bill   tip   sex smoker  day    time  size  price_per_person  \
156       48.17  5.00  Male     No  Sun  Dinner     6              8.03   
235       10.07  1.25  Male     No  Sat  Dinner     2              5.04   

        Payer Name         CC Number Payment ID  
156  Ryan Gonzales  3523151482063321    Sun7518  
235  Sean Gonzalez  3534021246117605    Sat4615

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

In [41]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [42]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [43]:
df.head(3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458

In [44]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [46]:
len(df)

244

In [47]:
df.shape

(244, 11)

In [48]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [49]:
df.sample()

total_bill  tip   sex smoker   day   time  size  price_per_person  \
138        16.0  2.0  Male    Yes  Thur  Lunch     2               8.0   

        Payer Name         CC Number Payment ID  
138  Jason Burgess  3561461821942363   Thur2710

In [50]:
df.aggregate

<bound method DataFrame.aggregate of      total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name   

In [51]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [52]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [53]:
df[df['total_bill']>48].sort_values(by='total_bill',ascending=False)

total_bill    tip   sex smoker  day    time  size  price_per_person  \
170       50.81  10.00  Male    Yes  Sat  Dinner     3             16.94   
212       48.33   9.00  Male     No  Sat  Dinner     4             12.08   
59        48.27   6.73  Male     No  Sat  Dinner     4             12.07   
156       48.17   5.00  Male     No  Sun  Dinner     6              8.03   

          Payer Name         CC Number Payment ID  
170    Gregory Clark  5473850968388236    Sat1954  
212  Alex Williamson      676218815212    Sat4590  
59       Brian Ortiz  6596453823950595    Sat8139  
156    Ryan Gonzales  3523151482063321    Sun7518

In [54]:
df[df['total_bill']>=46.13]

total_bill    tip   sex smoker  day    time  size  price_per_person  \
59        48.27   6.73  Male     No  Sat  Dinner     4             12.07   
156       48.17   5.00  Male     No  Sun  Dinner     6              8.03   
170       50.81  10.00  Male    Yes  Sat  Dinner     3             16.94   
212       48.33   9.00  Male     No  Sat  Dinner     4             12.08   

          Payer Name         CC Number Payment ID  
59       Brian Ortiz  6596453823950595    Sat8139  
156    Ryan Gonzales  3523151482063321    Sun7518  
170    Gregory Clark  5473850968388236    Sat1954  
212  Alex Williamson      676218815212    Sat4590

In [36]:
df['time'].value_counts()

time
Dinner    176
Lunch      68
Name: count, dtype: int64

In [37]:
df['sex'].value_counts()

sex
Male      157
Female     87
Name: count, dtype: int64

#### Grab Multiple Columns

In [31]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip','day']].sort_values(by='total_bill',ascending=False)

total_bill    tip   day
170       50.81  10.00   Sat
212       48.33   9.00   Sat
59        48.27   6.73   Sat
156       48.17   5.00   Sun
182       45.35   3.50   Sun
..          ...    ...   ...
149        7.51   2.00  Thur
111        7.25   1.00   Sat
172        7.25   5.15   Sun
92         5.75   1.00   Fri
67         3.07   1.00   Sat

[244 rows x 3 columns]

#### Create New Columns

In [10]:
df['tip_percentage'] = 100* (df['tip'] / df['total_bill'])

In [11]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [12]:
df['price_per_person'] = df['total_bill'] / df['size']

In [13]:
df['price_per_person'] = np.round(df['price_per_person'],2)

In [14]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [38]:
df[df['Payer Name']=='Douglas Tucker']

total_bill   tip   sex smoker  day    time  size  price_per_person  \
1       10.34  1.66  Male     No  Sun  Dinner     3              3.45   

       Payer Name         CC Number Payment ID  tip_percentage  
1  Douglas Tucker  4478071379779230    Sun4608       16.054159

In [39]:
help(np.round)

Help on _ArrayFunctionDispatcher in module numpy:

round(a, decimals=0, out=None)
    Evenly round to the given number of decimals.
    
    Parameters
    ----------
    a : array_like
        Input data.
    decimals : int, optional
        Number of decimal places to round to (default: 0).  If
        decimals is negative, it specifies the number of positions to
        the left of the decimal point.
    out : ndarray, optional
        Alternative output array in which to place the result. It must have
        the same shape as the expected output, but the type of the output
        values will be cast if necessary. See :ref:`ufuncs-output-type` for more
        details.
    
    Returns
    -------
    rounded_array : ndarray
        An array of the same type as `a`, containing the rounded values.
        Unless `out` was specified, a new array is created.  A reference to
        the result is returned.
    
        The real and imaginary parts of complex numbers are rounded
      

#### Adjust Existing Columns

In [40]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [41]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Remove Columns

In [42]:
# df.drop('tip_percentage',axis=1)

In [15]:
df = df.drop("tip_percentage",axis=1)

In [16]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [45]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [48]:
df.index

Index(['Sun2959', 'Sun4608', 'Sun4458', 'Sun5260', 'Sun2251', 'Sun9679',
       'Sun5985', 'Sun8157', 'Sun6820', 'Sun3775',
       ...
       'Sat7220', 'Sat4615', 'Sat5032', 'Sat2929', 'Sat9777', 'Sat2657',
       'Sat1766', 'Sat3880', 'Sat17', 'Thur672'],
      dtype='object', name='Payment ID', length=244)

In [17]:
df = df.set_index('Payment ID')

In [18]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [50]:
df = df.set_index('Payment ID')

In [51]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959             8.495000  Christy Cunningham  3560325168603410  
Sun4608             3.446667      Douglas Tucker  4478071379779230  
Sun4458             7.003333      Travis Walters  6011812112971322  
Sun5260            11.840000    Nathaniel Harris  4676137647685994  
Sun2251             6.147500        Tonya Carter  4832732618637221

In [52]:
df = df.reset_index()

In [53]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0          8.495000  Christy Cunningham  3560325168603410  
1          3.446667      Douglas Tucker  4478071379779230  
2          7.003333      Travis Walters  6011812112971322  
3         11.840000    Nathaniel Harris  4676137647685994  
4          6.147500        Tonya Carter  4832732618637221

### ROWS

Let's now explore these same concepts but with Rows.

In [54]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0          8.495000  Christy Cunningham  3560325168603410  
1          3.446667      Douglas Tucker  4478071379779230  
2          7.003333      Travis Walters  6011812112971322  
3         11.840000    Nathaniel Harris  4676137647685994  
4          6.147500        Tonya Carter  4832732618637221

In [55]:
df = df.set_index('Payment ID')

In [56]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959             8.495000  Christy Cunningham  3560325168603410  
Sun4608             3.446667      Douglas Tucker  4478071379779230  
Sun4458             7.003333      Travis Walters  6011812112971322  
Sun5260            11.840000    Nathaniel Harris  4676137647685994  
Sun2251             6.147500        Tonya Carter  4832732618637221

#### Grab a Single Row

In [50]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [19]:
df.iloc[232]

total_bill                     11.61
tip                             3.39
sex                             Male
smoker                            No
day                              Sat
time                          Dinner
size                               2
price_per_person                 5.8
Payer Name              James Taylor
CC Number           6011482917327995
Name: Sat2124, dtype: object

In [51]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [52]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [53]:
df.loc[['Sun5260','Sun4608']]

total_bill   tip   sex smoker  day    time  size  \
Payment ID                                                     
Sun5260          23.68  3.31  Male     No  Sun  Dinner     2   
Sun4608          10.34  1.66  Male     No  Sun  Dinner     3   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun4608                 3.45    Douglas Tucker  4478071379779230

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [54]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [55]:
df.drop('sex',axis=1).head()

total_bill   tip smoker  day    time  size  price_per_person  \
Payment ID                                                                 
Sun2959          16.99  1.01     No  Sun  Dinner     2              8.49   
Sun4608          10.34  1.66     No  Sun  Dinner     3              3.45   
Sun4458          21.01  3.50     No  Sun  Dinner     3              7.00   
Sun5260          23.68  3.31     No  Sun  Dinner     2             11.84   
Sun2251          24.59  3.61     No  Sun  Dinner     4              6.15   

                    Payer Name         CC Number  
Payment ID                                        
Sun2959     Christy Cunningham  3560325168603410  
Sun4608         Douglas Tucker  4478071379779230  
Sun4458         Travis Walters  6011812112971322  
Sun5260       Nathaniel Harris  4676137647685994  
Sun2251           Tonya Carter  4832732618637221

In [56]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [57]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [82]:
one_row = df.iloc[0]

In [83]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [84]:
type(one_row)

pandas.core.series.Series

In [85]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  
Payment ID                                                       
Sat2657                 9.68    Michael Avila  5296068606052842  
Sat1766                13.59   Monica Sanders  3506806155565404  
Sat3880                11.34       Keith Wong  6011891618747196  
Sat17                   8.91     Dennis Dixon     4375220550950  
Thur672                 9.39  Michelle Hardin  3511451626698139

In [20]:
df.append(one_row).tail()

AttributeError: 'DataFrame' object has no attribute 'append'

--------